src/Sub_internal.re

```diff
@@ -0,0 +1,89 @@
+module type Config = {
+  type params;
+  type msg;
+
+  // State that is carried by the subscription while it is active
+  type state;
+
+  let subscriptionName: string;
+
+  let getUniqueId: params => string;
+
+  let init: (~params: params, ~dispatch: msg => unit) => state;
+  let update:
+    (~params: params, ~state: state, ~dispatch: msg => unit) => state;
+  let dispose: (~params: params, ~state: state) => unit;
+};
+
+type config('params, 'msg, 'state) = (module Config with
+                                         type msg = 'msg and
+                                         type params = 'params and
+                                         type state = 'state);
+
+type subscription('params, 'msg, 'state) = {
+  config: config('params, 'msg, 'state),
+  params: 'params,
+  state: option('state),
+  // This is the same trick used in ReactMini -
+  // used to avoid Obj.magic when updating subscriptions.
+  handedOffInstance: ref(option('state)),
+};
+
+type t('msg) =
+  | NoSubscription: t('msg)
+  | Subscription(
+      subscription('params, 'originalMsg, 'state),
+      'originalMsg => 'msg,
```

Comment:
This shouldn't be necessary either. Replacing the first-class module with a record would make it possible to map the dispatch functions directly.

Reply:
Definitely. I didn't think about this until you mentioned it on Monday - I think that would be the next improvement of the API.

I believe this would enable polymorphic message types instead of needing to be baked into the functor, which would be a big improvement for the API!